File: 2019/python/day5.py
R, W = 0, 1

parms = {
    1: (R, R, W), 2: (R, R, W), 3: (W,), 4: (R,),
    5: (R, R), 6: (R, R), 7: (R, R, W), 8: (R, R, W)
}

def run(pgm, inp):
    p = list(pgm)
    output = None
    ip = 0

    while p[ip] != 99:
        modes, opcode = divmod(p[ip], 100)
        ip += 1

        args = [0, 0, 0]
        for n, rw in enumerate(parms[opcode]):
            val, ip = p[ip], ip + 1
            modes, m = divmod(modes, 10)
            if rw == R and m == 0:
                val = p[val]
            args[n] = val

        a, b, c = args
        if opcode == 1:
            p[c] = a + b
        elif opcode == 2:
            p[c] = a * b
        elif opcode == 3:
            p[a] = inp
        elif opcode == 4:
            print(a)
            output = a
        elif opcode == 5:
            if a != 0:
                ip = b
        elif opcode == 6:
            if a == 0:
                ip = b
        elif opcode == 7:
            p[c] = 1 if a < b else 0
        elif opcode == 8:
            p[c] = 1 if a == b else 0

    return output

with open('../day_5.txt', 'r') as f:
    pgm = [int(s) for s in f.read().rstrip().split(',')]

print('part 1:', run(pgm, 1))
print('part 2:', run(pgm, 5))

File: 2019/go/day5.go
package main

import (
	"AdventOfCode/intcode"
	"AdventOfCode/util"
	"fmt"
)

func part1(pgm string) int {
	computer, _ := intcode.LoadProgram(pgm)
	computer.Verbose(true)
	computer.Write(1)
	computer.Run()
	return computer.Read()
}

func part2(pgm string) int {
	computer, _ := intcode.LoadProgram(pgm)
	computer.Write(5)
	computer.Run()
	return computer.Read()
}

func main() {
	input, _ := util.ReadInput("../day_5.txt")
	pgm := input[0]
	fmt.Printf("part 1: %d\n", part1(pgm))
	fmt.Printf("part 2: %d\n", part2(pgm))
}
